src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+#nullable enable
+namespace Microsoft.CodeAnalysis
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0016:Add public types and members to the declared API", Justification = "In Progress")]
+    public abstract class GeneratorDriver
+    {
+        protected readonly GeneratorDriverState _state;
+
+        protected GeneratorDriver(GeneratorDriverState state)
+        {
+            _state = state;
+        }
+
+        protected GeneratorDriver(Compilation compilation, ParseOptions parseOptions)
+        {
+            _state = new GeneratorDriverState(compilation, parseOptions, ImmutableArray<GeneratorProvider>.Empty, ImmutableArray<AdditionalText>.Empty, ImmutableArray<PendingEdit>.Empty, ImmutableDictionary<GeneratorProvider, ImmutableArray<GeneratedSourceText>>.Empty, finalCompilation: null, editsFailed: true);
+        }
+
+        public GeneratorDriver GenerateSource(Compilation compilation, out Compilation outputCompilation, CancellationToken cancellationToken = default)
+        {
+            // with no providers, there is no work to do
+            if (_state.Providers.Length == 0)
+            {
+                outputCompilation = compilation;
+                return this;
+            }
+
+            // PERF: if the input compilation is the same as the last compilation we saw, and we have a final compilation
+            //       we know nothing can have changed and can just short circuit, returning the already processed final compilation 
+            if (compilation == _state.Compilation && _state.FinalCompilation is object)
+            {
+                outputCompilation = _state.FinalCompilation;
+                return this;
+            }
+
+            // run the actual generation
+            var state = StateWithPendingEditsApplied(_state);
+            var sourcesBuilder = PooledDictionary<GeneratorProvider, ImmutableArray<GeneratedSourceText>>.GetInstance();
+            foreach (var provider in state.Providers)
```

Comment:
Do we want to consider running this in parallel?

Reply:
Yup, I actually thought I'd added a proto comment for it. Added.

---
In reply to: [381649938](https://github.com/dotnet/roslyn/pull/41789#discussion_r381649938) [](ancestors = 381649938)